### DISCUSSION TITLE
sympy wrapper to verify derivation of simple algebraic expressions

### DISCUSSION THREAD

**Discussion by Des-Nerger** [sympy/sympy#24899] [Q&A] [answered]
I often derive algebraic formulas like this one:
```
2*(2*(i+0.5)-(j+0.5)) = 2*(2*i+1-j-0.5) = 4*i+2-2*j-1 = 4*i-2*j+1
```
writing such derivations in a text file.
Is there some kind of wrapper around sympy to verify with minimal hassle if those derivations are free of mistakes?

**Comment by oscarbenjamin** [comment, marked as answer]
You could use something like this:
```python
In [6]: s = '2*(2*(i+0.5)-(j+0.5)) = 2*(2*i+1-j-0.5) = 4*i+2-2*j-1 = 4*i-2*j+1'

In [7]: exprs = [nsimplify(parse_expr(si)) for si in s.split('=')]

In [8]: exprs
Out[8]: [4⋅i - 2⋅j + 1, 4⋅i - 2⋅j + 1, 4⋅i - 2⋅j + 1, 4⋅i - 2⋅j + 1]

In [9]: [simplify(e - exprs[0]) for e in exprs]
Out[9]: [0, 0, 0, 0]

In [10]: all(simplify(e - exprs[0]) == 0 for e in exprs)
Out[10]: True
```
More efficient approaches can be possible if you have constraints on the form of the expressions e.g. that they are always polynomial etc. It is not guaranteed that `simplify` will always prove an expression equal to zero so this approach could throw up some false positives (i.e. an expression that is valid might not be proven valid) but should not throw up false negatives.
